compiler/decls.go

```diff
@@ -0,0 +1,541 @@
+	}
+
+	var varDecls []*Decl
+	varsWithInit := fc.pkgCtx.VarsWithInitializers()
+
+	initializers := []*types.Initializer{}
+
+	// For implicitly-initialized vars we generate synthetic zero-value
+	// initializers and then process them the same way as explicitly initialized.
+	for _, o := range vars {
+		if varsWithInit[o] {
+			continue
+		}
+		initializer := &types.Initializer{
+			Lhs: []*types.Var{o},
+			Rhs: fc.zeroValue(o.Type()),
+		}
+		initializers = append(initializers, initializer)
+	}
+
+	// Add explicitly-initialized variables to the list. Implicitly-initialized
+	// variables should be declared first in case explicit initializers depend on
+	// them.
+	initializers = append(initializers, fc.pkgCtx.InitOrder...)
+
+	for _, init := range initializers {
+		varDecls = append(varDecls, fc.newVarDecl(init))
+	}
+
+	return varDecls
+}
+
+// newVarDecl creates a new Decl describing a variable, given an explicit
+// initializer.
+func (fc *funcContext) newVarDecl(init *types.Initializer) *Decl {
+	var d Decl
+
+	assignLHS := []ast.Expr{}
+	for _, o := range init.Lhs {
+		assignLHS = append(assignLHS, fc.newIdentFor(o))
+
+		// For non-exported package-level variables we need to declared a local JS
+		// variable. Exported variables are represented as properties of the $pkg
+		// JS object.
+		if !o.Exported() {
+			d.Vars = append(d.Vars, fc.objectName(o))
+		}
+		if fc.pkgCtx.HasPointer[o] && !o.Exported() {
+			d.Vars = append(d.Vars, fc.varPtrName(o))
+		}
+	}
+
+	d.DceDeps = fc.CollectDCEDeps(func() {
+		fc.localVars = nil
+		d.InitCode = fc.CatchOutput(1, func() {
+			fc.translateStmt(&ast.AssignStmt{
+				Lhs: assignLHS,
+				Tok: token.DEFINE,
+				Rhs: []ast.Expr{init.Rhs},
+			}, nil)
+		})
+
+		// Initializer code may have introduced auxiliary variables (e.g. for
+		// handling multi-assignment or blocking calls), add them to the decl too.
+		d.Vars = append(d.Vars, fc.localVars...)
+		fc.localVars = nil // Clean up after ourselves.
+	})
+
+	if len(init.Lhs) == 1 {
+		if !analysis.HasSideEffect(init.Rhs, fc.pkgCtx.Info.Info) {
+			d.DceObjectFilter = init.Lhs[0].Name()
+		}
+	}
+	return &d
+}
+
+// funcDecls translates all package-level function and methods.
+//
+// `functions` must contain all package-level function and method declarations
+// found in the AST. The function returns Decls that define corresponding JS
+// functions at runtime. For special functions like init() and main() decls will
+// also contain code necessary to invoke them.
+func (fc *funcContext) funcDecls(functions []*ast.FuncDecl) ([]*Decl, error) {
+	if !fc.IsRoot() {
+		panic(bailout(fmt.Errorf("functionContext.funcDecls() must be only called on the package-level context")))
+	}
+
+	var mainFunc *types.Func
+	funcDecls := []*Decl{}
+	for _, fun := range functions {
+		funcDecls = append(funcDecls, fc.newFuncDecl(fun))
+
+		if o := fc.pkgCtx.Defs[fun.Name].(*types.Func); o.Name() == "main" {
+			mainFunc = o // main() function candidate.
```

Comment:
"candidate"?  Can there be more than one?

Reply:
Random side note: Is there any compile-time check that `main()` takes no arguments, and has no return values?

Reply:
Both are good questions...

I *think* there can't be two main in a single package functions in a type-checked code. There *can* be multiple candidates in the whole program though, e.g. if your main package has tests, `gopherjs test` would synthesize another main function. I am actually unsure what happens then, the current implementation looks like we'd invoke both main functions, which is clearly wrong.

Reply:
It seems the check is for make's arguments is being done, just not by us.

https://gopherjs.github.io/playground/#/2smtCKvpi_

I haven't tested the test scenario you describe, but I expect it would have bitten me in the past if it were a problem.  When I have some time, I'll try to verify, and file a bug if it's an issue.